lib/passport-azure-ad/oidcstrategy.js

```diff
@@ -0,0 +1,203 @@
+ *     passport.use(new OIDCBearerStrategy(
+ *       secretOrPublicKey
+ *       function(token, done) {
+ *         User.findById(token.sub, function (err, user) {
+ *           if (err) { return done(err); }
+ *           if (!user) { return done(null, false); }
+ *           return done(null, user, token);
+ *         });
+ *       }
+ *     ));
+ *
+ * For further details on HTTP Bearer authentication, refer to [The OAuth 2.0 Authorization Protocol: Bearer Tokens](http://tools.ietf.org/html/draft-ietf-oauth-v2-bearer)
+ * For further details on JSON Web Token, refert to [JSON Web Token](http://tools.ietf.org/html/draft-ietf-oauth-json-web-token)
+ *
+ * @param {object} options - The Options.
+ * @param {Function} verify - The verify callback.
+ * @constructor
+ */
+
+function Strategy(options, verify) {
+
+
+  // if you want to check JWT issuer that is not in the token (not recommended), provide a value here
+
+  if (options.issuer) {
+    log.info('Issuer provided to Strategy was: ', options.issuer);
+  }
+
+  // if you want to check JWT audience (aud), provide a value here
+
+  if (options.audience) {
+    log.info('Audience provided to Strategy was: ', options.audience);
+  }
+
+  if (options.identityMetadata) {
+
+    log.info('Metadata url provided to Strategy was: ', options.identityMetadata);
+    this.metadata = new Metadata(options.identityMetadata, "oidc");
+  }
+
+  if (!options.certificate && !options.identityMetadata) {
+    log.warn("No options was presented to Strategy as required.");
+    throw new TypeError('OIDCBearerStrategy requires either a PEM encoded public key or a metadata location that contains cert data for RSA and ECDSA callback.');
+  }
+
+  if (typeof options === 'function') {
+    verify = options;
+    options = {};
+  }
+
+  // Passport requires a verify function
+
+  if (!verify) {
+    throw new TypeError('OIDCBearerStrategy requires a verify callback. Do not cheat!');
+  }
+
+  // Token validation settings. Hopefully most of these will be pulled from the metadata and this is not needed
+
+
+  this.metadata.fetch(function(err) {
+    if (err) {
+      throw new Error("Unable to fetch metadata: " + err);
+    }
+
+  });
+
+
+  function jwtVerify(req, token, done) {
+
+    if (!options.passReqToCallback) {
+      token = arguments[0];
+      done = arguments[1];
+      req = null;
+      log.info(token, 'was token going in to verification');
+    }
+
+
+    var decoded = jws.decode(token);
+    if (decoded == null) {
+      done(null, false, "Invalid JWT token.");
+    }
+
+    log.info('token decoded. But is it valid?: ', decoded);
+
+
+    // We have two different types of token signatures we have to validate here. One provides x5t and the other a kid.
+    // We need to call the right one.
+
+    if (decoded.header.x5t) {
+      PEMkey = this.metadata.generateOidcPEM(decoded.header.x5t);
+    } else if (decoded.header.kid) {
+      PEMkey = this.metadata.generateOidcPEM(decoded.header.kid);
+    } else {
```

Comment:
Here is where I go out of my depth - I thought that we either validated with x5t or kid. What would we use to validate if not x5t? Something the user would pass?